[Select]: New Select component (Beta)

AbhinavMV opened the pull request:
---
name: 'Pull request'
about: Select component
---

## New Pull Request

### Checklist

<!--
    Check every following box [x] before submitting your PR.
    Click the "Preview" tab for better readability.
    Thanks for contributing to Moralis!
-->

- [x] I am not disclosing a [vulnerability](https://github.com/MoralisWeb3/Moralis-JS-SDK/blob/main/SECURITY.md).
- [x] My code is conform the [code style](https://github.com/web3ui/web3uikit/blob/master/CONTRIBUTE.md)
- [x] I have made corresponding changes to the documentation
- [x] I have updated Typescript definitions when needed

### Issue Description

<!-- Add a brief description of the issue this PR solves. -->

Related issue: #749 , #728

### Solution Description

<!-- Add a description of the solution in this PR. -->
1. Folder structure changed 
2. New select component added in beta version since more changes need to be done in further iterations.
3. change `tryBeta` prop to true to try this component
4. Search feature added
5. Multi-selection added
6. Accessibility (Keyboard) added

BillyG83 commented:
wow this looks great @AbhinavMV i think i only have two TINY bit of feedback
- on change it would be nice to see the action in storybook, i dont even see the console log actually
- each name could use like 4px padding left on the span
- 2 countries flags fail, prob not your bad

![padding](https://user-images.githubusercontent.com/13779759/185091683-c26e587f-2e81-4303-9270-8fdd3dc826b2.png)

![broken flag](https://user-images.githubusercontent.com/13779759/185091688-8a2a8cc6-ea31-455e-80ad-9a4117c5ba26.png)

![broken flag 2](https://user-images.githubusercontent.com/13779759/185091690-7163fd99-b6e4-4639-8462-09f9b7f69ebb.png)

BillyG83 commented on .changeset/thirty-mayflies-reply.md:
```diff
@@ -0,0 +1,6 @@
+---
```
what is this file? 😅

AbhinavMV replied:
its for auto publishing of packages. (will be useful once it is fixed 😂)
all PR now need this file.

AbhinavMV replied:
> * on change it would be nice to see the action in storybook, i dont even see the console log actually

Added action for this 🙌🏻 

> * each name could use like 4px padding left on the span
Fixed 🙌🏻 

> * 2 countries flags fail, prob not your bad
> 
> ![padding](https://user-images.githubusercontent.com/13779759/185091683-c26e587f-2e81-4303-9270-8fdd3dc826b2.png)
> 
> ![broken flag](https://user-images.githubusercontent.com/13779759/185091688-8a2a8cc6-ea31-455e-80ad-9a4117c5ba26.png)
> 
> ![broken flag 2](https://user-images.githubusercontent.com/13779759/185091690-7163fd99-b6e4-4639-8462-09f9b7f69ebb.png)
Yeah no image present for that, can't do anything 😅 

@BillyG83 I guess all the mentioned issues are fixed. 
Also new things added:
1. default index can be set now
2. Custom error and description messages

Validation is pending - probably add it in another PR? this is getting huge would make it hard to review 😅

BillyG83 commented on packages/core/src/lib/Select/OldSelect/OldSelect.styles.ts:
```diff
@@ -0,0 +1,237 @@
+import { color, fonts, resetCSS } from '@web3uikit/styles';
+import styled, { css } from 'styled-components';
+
+const DivStyledWrapper = styled.div<Pick<ISelectProps, 'state'>>`
```
it looks like you dont need `<Pick<ISelectProps, 'state'>>` here, right?

BillyG83 commented on packages/core/src/lib/Select/OldSelect/OldSelect.styles.ts:
```diff
@@ -0,0 +1,237 @@
+  position: absolute;
+  padding: 0 4px;
+  transition: all 0.1s ease-out;
+  ${({ hasSelectedIndex }) =>
```
weird Props name IMO, `isSelected` or `isActive` maybe?

BillyG83 commented on packages/core/src/lib/Select/OldSelect/OldSelect.styles.ts:
```diff
@@ -0,0 +1,237 @@
+`;
+
+const PrefixSpan = styled.span`
+    margin-right: px;
```
`margin-right: px;` i've never seen this, do you want `margin-right: 0;` maybe?

BillyG83 commented on packages/core/src/lib/Select/OldSelect/OldSelect.styles.ts:
```diff
@@ -0,0 +1,237 @@
+
+    border-color: ${({ state }) => {
+        switch (state) {
+            case 'error':
```
how can a Select be in error?

BillyG83 replied:
maybe if it is required but missed by the user 🤔

AbhinavMV replied:
it can have error state for validation
not yet added in the new version.
will do a bit more research for this 🙌

BillyG83 commented on packages/core/src/lib/Select/SelectBeta/Select.test.tsx:
```diff
@@ -0,0 +1,189 @@
+        // search is hidden test
+        expect(searchInput).toHaveAttribute('aria-hidden', 'true');
+    });
+});
```
nice tests, how did the coverage turn out?

AbhinavMV replied:
will check 🙌
will add a bit more tests 😋

AbhinavMV replied:
![image](https://user-images.githubusercontent.com/35369843/185473006-95c2ccac-b621-497e-a695-a4c94f0bc7f3.png)
Not bad, Needs more tests so that future modifications won't break this

BillyG83 commented on packages/core/src/lib/Select/SelectBeta/SelectBeta.tsx:
```diff
@@ -0,0 +1,182 @@
+    const triggerRef = useRef<HTMLButtonElement>(null);
+    const [isOpen, setIsOpen] = useState(false);
+
+    function elId(component: string) {
```
i'm not against `function() {}` but the rest or most of the kit uses `const functionName = () => {}`
its nice to try and align don't you think?

BillyG83 replied:
also `elId` sounds like a defined const to me, not the name of a function that will normalise a component ID, maybe consider making the name more friendly for the next dev 🤷‍♂️

AbhinavMV replied:
agreed 💯. 
my late night naming skills are not good 😂

AbhinavMV replied:
Changed the name to elementId (elId was short for this 😆) ✅

BillyG83 commented on packages/core/src/lib/Select/SelectBeta/SelectBeta.tsx:
```diff
@@ -0,0 +1,182 @@
+        if (!option) return;
+        // multi selection
+        if (isMulti === true) {
+            let selected = [...value];
```
i think this can be a const, right? you are making an array to mutate so i do not think you need a `let`

AbhinavMV replied:
aah right nice catch 🙌🔥

AbhinavMV replied:
This is fixed ✅

BillyG83 commented on packages/core/src/lib/Select/OldSelect/OldSelect.styles.ts:
```diff
@@ -0,0 +1,237 @@
+import { LabelProps, SelectedItemProps, ISelectProps } from '../types';
```
i just noticed this is OldSelect after leaving comments, so this is not your code, right?

AbhinavMV replied:
yup this is just the old component
have kept it for now. till we finalize the new version 😄
resolved all conversation that are on old component file 🙌

BillyG83 reviewed (commented):
this looks great dude, i think all my comments were on the old mode until i realised, feel free to close any that are not to your code.

AbhinavMV commented:
> wow this looks great @AbhinavMV i think i only have 3 TINY bits of feedback:
> 
> * on change it would be nice to see the action in storybook, i dont even see the console log actually
> * each name could use like 4px padding left on the span
> * 2 countries flags fail, prob not your bad
> 
> ![padding](https://user-images.githubusercontent.com/13779759/185091683-c26e587f-2e81-4303-9270-8fdd3dc826b2.png)
> 
> ![broken flag](https://user-images.githubusercontent.com/13779759/185091688-8a2a8cc6-ea31-455e-80ad-9a4117c5ba26.png)
> 
> ![broken flag 2](https://user-images.githubusercontent.com/13779759/185091690-7163fd99-b6e4-4639-8462-09f9b7f69ebb.png)

Nice catch.
Got it will fix all these soon. 🙌

AbhinavMV commented:
> * on change it would be nice to see the action in storybook, i dont even see the console log actually

Added action for this 🙌🏻 

> * each name could use like 4px padding left on the span

Fixed 🙌🏻 

> * 2 countries flags fail, prob not your bad
> 
> ![padding](https://user-images.githubusercontent.com/13779759/185091683-c26e587f-2e81-4303-9270-8fdd3dc826b2.png)
> 
> ![broken flag](https://user-images.githubusercontent.com/13779759/185091688-8a2a8cc6-ea31-455e-80ad-9a4117c5ba26.png)
> 
> ![broken flag 2](https://user-images.githubusercontent.com/13779759/185091690-7163fd99-b6e4-4639-8462-09f9b7f69ebb.png)

Yeah no image present for that, can't do anything 😅 

@BillyG83 I guess all the mentioned issues are fixed. 
Also new things added:
1. default index can be set now
2. Custom error and description messages

Validation is pending - probably add it in another PR? this is getting huge would make it hard to review 😅

BillyG83 commented:
great job @AbhinavMV lets merge this and test in Admin to see if QA will pass it, and get feedback from them. Also if you could look into a fallback image for any flags that are missing or fail for any reason, the broken image thing is a little ugly and @MateusbrMedeiros will not pass that I am sure 😁
